### ISSUE TITLE
FormGroup not dirty when a nested FormControl changes from disabled to enabled, or vice versa.

### ISSUE THREAD

**Issue by jackwootton** [angular/angular#34252]
# 🐞 bug report

### Affected Package
@angular/forms

### Is this a regression?
No

### Description
A [FormGroup](https://angular.io/api/forms/FormGroup) is not considered [dirty](https://angular.io/api/forms/AbstractControl#dirty) when a nested [FormControl](https://angular.io/api/forms/FormControl) is changed from disabled to enabled, or vice versa. The [pristine](https://angular.io/api/forms/AbstractControl#pristine), [dirty](https://angular.io/api/forms/AbstractControl#dirty), [touched](https://angular.io/api/forms/AbstractControl#touched), and [untouched](https://angular.io/api/forms/AbstractControl#untouched) properties remain unchanged when a control changes from [enabled](https://angular.io/api/forms/AbstractControl#enable) to [disabled](https://angular.io/api/forms/AbstractControl#disable), or vice versa. 

## 🔬 Minimal Reproduction
https://stackblitz.com/edit/formbuilder-6cjren

## 🌍  Your Environment

**Angular Version:**
<pre><code>
4.3.6
</code></pre>

**Anything else relevant?**
Posted more detail on StackOverflow: https://stackoverflow.com/questions/59184512/get-all-angular-form-controls-that-have-changed-from-enabled-to-disabled-or-vic

**Comment by pkozlowski-opensource** [comment]
@jackwootton Angular version 4.x is a couple of years old... Can you reproduce the same issue with the latest (8.x) version?

**Comment by jackwootton** [comment]
@pkozlowski-opensource - I've upgraded the Stackblitz to `8.0.0`. Looks like it's reproducible.

**Comment by vladjerca** [comment]
@jackwootton I don't see why a form should become dirty if the `disabled` state of an input changes. Could you explain why that should be the case? I personally don't get it. Conceptually I see a form should be dirty if the user modifies its value. 

Normally a disabled input means that the data from that input will be omitted from the form value.

So let's say I have a prepopulated form, and I enable the input, that only means that I now want to include that value as part of the form data, but I haven't modified it's value yet. So the data is still `pristine` aka `not modified by the user`.

**Comment by jackwootton** [comment]
@vladjerca The value _does_ change. In [my Stackblitz](https://stackblitz.com/edit/formbuilder-6cjren) the `value` of the form is printed:

Enabled:
```
{
  "name": "",
  "age": 100
}
```

Disabled:
```
{
  "age": 100
}
```

The value of the FormGroup is clearly different. I don't see that the _means_ of changing the value is relevant in determining whether the value of the FormGroup has changed from its original state.

**Comment by jackwootton** [comment]
Perhaps this should be a feature request?

- Add a property to FormGroup reflecting whether any controls have been toggled away from their original enable/disabled state.
_or_
- Add a property to FormControl reflecting whether the control has been toggled away from its original enable/disabled state.

**Comment by wanderrful** [comment]
@jackwootton Hey I spent some time looking into this for you and found this comment on `packages/forms/src/model.ts` L554:

```
// If parent has been marked artificially dirty we don't want to re-calculate the
// parent's dirtiness based on the children.
```

I also found this old PR to try and fix the issue you're talking about in your post and it looks like the Angular folks specifically didn't want it to work this way (at least at the time):

https://github.com/angular/angular/pull/24591

Near the bottom of that page, you can see the commit where they added that comment I quoted above.

**Comment by jackwootton** [comment]
@wanderrful Thanks for making the time to look into this. After reading [#24591](https://github.com/angular/angular/pull/24591) I think I've conflated a few issues. Let me tease a few things apart;

1. A feature request for a new flag, `disabledChanged` for example, which is `true` when the disabled status has changed. On a `FormGroup`, this flag would reflect whether the disabled status of any nested controls have changed. On a `FormControl` this flag would would reflect whether the disabled status of said control has changed.

2. [#24591](https://github.com/angular/angular/pull/24591)

3. Similar to [#24591](https://github.com/angular/angular/pull/24591), but without explicitly marking the control as dirty, i.e. 

```typescript
form = this.fb.group({ a: ['a'], b: ['b'] });
form.controls['a'].disable();
console.log(form.dirty); // false
```

My reason for the third suggestion, is that the value of the `FormGroup` has changed from `{ "a": "a", "b": "b" }` to `{"b": "b" }`. However, I also agree with [#24591](https://github.com/angular/angular/pull/24591).

In summary: I think 1 and 2 are needed, then 3. becomes redundant. The current situation means it's necessary to maintain a disabled-changed state:

```typescript
private disabledChanged: { [name: string]: boolean; } = {
    // FormControl names.
    'a': false,
    'b': false,
};

public onDisabledChanged(name: string) {
    this.disabledChanged[name] = !this.disabledChanged[name];
}
```

**Comment by vladjerca** [comment]
> @vladjerca The value _does_ change. In [my Stackblitz](https://stackblitz.com/edit/formbuilder-6cjren) the `value` of the form is printed:
> 
> Enabled:
> 
> ```json
> {
>   "name": "",
>   "age": 100
> }
> ```
> 
> Disabled:
> 
> ```json
> {
>   "age": 100
> }
> ```
> 
> The value of the FormGroup is clearly different. It is dirty. I don't see that the _means_ of changing the value is relevant in determining whether the value of the FormGroup has changed from its original state.
> 
> ```js
> // false
> { "name": "", "age": 100 } === { "age": 100 }
> ```

I get your point, but somehow I still have mixed feelings about this. I still kind of see this as model `schema` change rather than the value becoming `dirty`.

But I guess augmenting the control with an original disabled state property would be helpful for a lot of use cases.

**Comment by jackwootton** [comment]
@vladjerca - I think you're right here. It does feel like more of a schema change. Perhaps accessing the `value` of a FormGroup threw me little.  A new `disabedStateChanged` would pretty much negate the need for the changing the meaning of `value` since it would become trivial to identify controls that have caused a 'schema` change.
